middle_end/closure/closure.ml

```diff
@@ -977,6 +977,34 @@ let close_approx_var { fenv; cenv } id =
 let close_var env id =
   let (ulam, _app) = close_approx_var env id in ulam
 
+let rec compute_expr_layout kinds lam =
+  match lam with
+  | Lvar id | Lmutvar id -> V.Map.find id kinds
+  | Lconst cst -> structured_constant_layout cst
+  | Lfunction _ -> Lambda.layout_function
+  | Lapply { ap_result_layout; _ } -> ap_result_layout
+  | Lsend (_, _, _, _, _, _, _, layout) -> layout
+  | Llet(_, kind, id, _, body) | Lmutlet(kind, id, _, body) ->
+    compute_expr_layout (V.Map.add id kind kinds) body
+  | Lletrec(defs, body) ->
+    let kinds =
+      List.fold_left (fun kinds (id, _) -> V.Map.add id Lambda.layout_letrec kinds)
+        kinds defs
+    in
+    compute_expr_layout kinds body
+  | Lprim(p, _, _) ->
+    Lambda.primitive_result_layout p
+  | Lswitch(_, _, _, kind) | Lstringswitch(_, _, _, _, kind)
+  | Lstaticcatch(_, _, _, kind) | Ltrywith(_, _, _, kind)
+  | Lifthenelse(_, _, _, kind) | Lregion (_, kind) ->
+    kind
+  | Lstaticraise (_, _) ->
+    Lambda.layout_top (* Need a layout_bottom here *)
```

Comment:
Maybe just declare a layout_bottom in lambda and use it here. Or wait for #1115 which declares it